src/metadataExt.ts

```diff
@@ -0,0 +1,261 @@
+    ): Promise<any[]> {
+        return (await this.metadataModule.getObjectUsing(projectID, dashboardUri, options)).filter(
+            (obj: any) => obj.category.match("kpi|visualizationWidget|filterContext"),
+        );
+    }
+
+    private async updateWidgets(
+        projectID: string,
+        dashboard: any,
+        dashboardID: string,
+        objectsUsingDashboard: any[],
+    ): Promise<string> {
+        const widgetLinks: string[] = objectsUsingDashboard
+            .filter((obj: any) => obj.category.match("kpi|visualizationWidget"))
+            .map((obj: any) => {
+                return obj.link;
+            });
+
+        const widgets: any[] = await Promise.all(
+            (await this.metadataModule.getObjects(projectID, widgetLinks)).map(async (widget: any) => {
+                const category = Object.keys(widget)[0];
+                if (category === "visualizationWidget") {
+                    const visualizationWidget = await this.metadataModule.createObject(projectID, widget);
+                    await this.updateVisWidget(projectID, visualizationWidget);
+                    return visualizationWidget;
+                }
+                return this.metadataModule.createObject(projectID, widget);
+            }),
+        );
+
+        const replaceWidgetsRef = Promise.all(
+            dashboard.content.widgets.map(async (widget: string) => {
+                return widgetLinks
+                    .map((link, index) => {
+                        if (link === widget) {
+                            if (Object.keys(widgets[index])[0] === "kpi") {
+                                const { kpi } = widgets[index];
+
+                                return kpi.meta.uri;
+                            } else {
+                                const { visualizationWidget } = widgets[index];
+
+                                return visualizationWidget.meta.uri;
+                            }
+                        }
+                    })
+                    .filter(links => links !== undefined)
+                    .join(",");
+            }),
+        );
+
+        const clonedLayout = this.updateLayout(dashboard, widgetLinks, widgets);
+
+        const updatedDashboardLink: string = await this.updateFilterContext(
+            projectID,
+            dashboardID,
+            dashboard,
+            objectsUsingDashboard,
+        );
+        const { analyticalDashboard } = await this.metadataModule.getObjectDetails(updatedDashboardLink);
+
+        const { uri }: { uri: string } = await this.metadataModule.updateObject(projectID, dashboardID, {
+            analyticalDashboard: {
+                ...analyticalDashboard,
+                content: {
+                    ...analyticalDashboard.content,
+                    layout: { ...analyticalDashboard.content.layout, fluidLayout: clonedLayout },
+                    widgets: await replaceWidgetsRef,
+                },
+            },
+        });
```

Comment:
this will not be needed once saveAs creates new dashboard object at the very end of processing. until then you can pass around the copy of analytical dashboard, keep updating the links in it; returning the updated object after each function